// rosie-skills/dist/index.js
// rosie-skills — typed JS API.
//
// Each function maps to a CLI subcommand. Options correspond 1:1 with CLI
// flags. By default the library is silent; failures throw with a descriptive
// message. Pass `onLog` to observe progress.
//
// Example:
//   import * as rosie from 'rosie-skills';
//   await rosie.install('anthropics/skills');
//   const skills = await rosie.list();
import { callApi, loadModule } from "./wasm-loader.js";
// Wrap a call with process.chdir to the requested cwd, restoring on exit.
async function withCwd(cwd, fn) {
    if (!cwd)
        return fn();
    const orig = process.cwd();
    process.chdir(cwd);
    try {
        return await fn();
    }
    finally {
        process.chdir(orig);
    }
}
export async function list(opts = {}) {
    return withCwd(opts.cwd, async () => {
        const mod = await loadModule(opts.onLog);
        return callApi(mod, "rosie_api_list_installed");
    });
}
export async function agents(opts = {}) {
    return withCwd(opts.cwd, async () => {
        const mod = await loadModule(opts.onLog);
        return callApi(mod, "rosie_api_agents");
    });
}
/**
 * Install a skill or reference. With no `spec`, reinstalls everything in
 * `.agents/rosie.lock` (matches the CLI's `rosie install` with no args).
 */
export async function install(spec, opts = {}) {
    return withCwd(opts.cwd, async () => {
        const mod = await loadModule(opts.onLog);
        const agents = Array.isArray(opts.agent) ? opts.agent.join(",") : opts.agent ?? "";
        const includes = (opts.include ?? []).join("\n");
        const skipLockfile = opts.lockfile === false ? 1 : 0;
        return await callApi(mod, "rosie_api_install", [
            spec,
            opts.skill ?? "",
            agents,
            opts.name ?? "",
            includes,
            opts.ref ? 1 : 0,
            opts.npm ? 1 : 0,
            opts.global ? 1 : 0,
            skipLockfile,
        ]);
    });
}
/** Reinstall everything from `.agents/rosie.lock`. */
export async function installFromLockfile(opts = {}) {
    return install("", opts);
}
export async function remove(skillName, opts = {}) {
    return withCwd(opts.cwd, async () => {
        const mod = await loadModule(opts.onLog);
        const agents = Array.isArray(opts.agent) ? opts.agent.join(",") : opts.agent ?? "";
        const skipLockfile = opts.lockfile === false ? 1 : 0;
        await callApi(mod, "rosie_api_remove", [
            skillName,
            agents,
            opts.global ? 1 : 0,
            skipLockfile,
        ]);
    });
}
/**
 * Update one skill (by name) or all entries if no name is given. Returns
 * the same `InstallResult` shape `install` returns — `skills` covers every
 * entry that was re-resolved (including those that ended up unchanged).
 */
export async function update(skillName, opts = {}) {
    return withCwd(opts.cwd, async () => {
        const mod = await loadModule(opts.onLog);
        const skipLockfile = opts.lockfile === false ? 1 : 0;
        return await callApi(mod, "rosie_api_update", [
            skillName ?? "",
            skipLockfile,
        ]);
    });
}
