// rosie-skills/dist/wasm-loader.js
// Loads the inlined WASM build and exposes a typed `RosieModule` interface.
// All API entry points come from wasm/rosie.js (an ESM module emitted by
// emcc with EXPORT_ES6=1); this module instantiates that and wires up
// the log-callback bridge.
//
// Logging: rosie's C log_* functions route through a registered callback;
// when API mode is active that callback bridges into JS via a Module slot
// (`__rosieLog__`). We can't use Module.print here because NODERAWFS routes
// printf to host fd 1/2 directly, bypassing the Module hook.
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { silenceWasiExperimentalWarning } from "./silence-wasi-warning.js";
const __dirname = path.dirname(fileURLToPath(import.meta.url));
// Internal: integer levels match the C-side LogLevel enum.
const LEVEL_NAMES = ["error", "warn", "info", "debug"];
let modulePromise = null;
// Load the module once per process and install the log bridge.
async function getOrLoadModule() {
    if (modulePromise)
        return modulePromise;
    silenceWasiExperimentalWarning();
    const wasmEntry = path.join(__dirname, "..", "wasm", "rosie.js");
    const mod = (await import(wasmEntry));
    const createRosie = mod.default;
    modulePromise = (async () => {
        const m = await createRosie({
            // Skip main(); API entry points are called via ccall.
            noInitialRun: true,
            // Module.print won't catch printf under NODERAWFS, so leave these as
            // defaults (they print to console.log/console.error). All rosie output
            // goes through log_* which we re-route via the bridge.
        });
        // Install the C-side log bridge so log_* messages reach __rosieLog__.
        m.ccall("rosie_api_install_log_bridge", null, [], []);
        // Tell C what host platform we're on so link.c can route symlink calls
        // through wasm_create_junction / wasm_copy_or_link_file on Windows.
        m.ccall("rosie_api_set_host_platform", null, ["string"], [process.platform]);
        return m;
    })();
    return modulePromise;
}
// Public: load the module and stash the per-call onLog handler on it before
// the C code runs. Returns the module so the caller can ccall API entries.
export async function loadModule(onLog) {
    const m = await getOrLoadModule();
    m.__rosieLog__ = onLog
        ? (level, message) => {
            const lvl = LEVEL_NAMES[level] ?? "info";
            onLog({ level: lvl, message });
        }
        : null;
    return m;
}
// Calls a C API function that returns a malloc'd char* JSON string. Copies
// the string out and frees the WASM-side buffer. Returns the parsed JSON.
//
// Always uses `async: true` so Asyncify-yielding ops (install, update — which
// call into JS fetch) work transparently. For non-yielding ops (list, agents,
// remove) the returned Promise just resolves on the same microtask.
export async function callApi(mod, fn, args = []) {
    const argTypes = args.map((a) => (typeof a === "number" ? "number" : "string"));
    const ptr = (await mod.ccall(fn, "number", argTypes, args, { async: true }));
    if (!ptr)
        throw new Error(`rosie-skills: ${fn} returned null`);
    try {
        const json = mod.UTF8ToString(ptr);
        return parseEnvelope(json);
    }
    finally {
        mod._free(ptr);
    }
}
function parseEnvelope(json) {
    let parsed;
    try {
        parsed = JSON.parse(json);
    }
    catch (e) {
        throw new Error(`rosie-skills: failed to parse WASM response: ${e.message}`);
    }
    if (!parsed.ok) {
        throw new Error(parsed.error || "rosie-skills: unknown error");
    }
    return parsed.data;
}
